### ISSUE TITLE
"Format Code" did the import wrong

### ISSUE THREAD

**Issue by mholt** [Microsoft/vscode-go#16]
I added a `fmt.Println()` to a file, which should have added "fmt" to the imports. When I right-clicked "Format Code" it did this instead:

```
import (fmt"io/ioutil"
    "log"
    "os"
    "os/signal"
    "syscall"
)
```

and kind of mashed the "fmt" improperly into the beginning of the list.

Did I do it wrong? :smile: Also, can it run gofmt on save?

**Comment by Gys** [comment]
Did you enable File -> Auto save ?
Because formatting this way works for me (using the default goreturns).

But formatting at save does not work for me at all. See other issue.

**Comment by mholt** [comment]
I did enable it, but I can't remember if I enabled it before or after this. I know in this particular instance, I right-clicked, then clicked "Format Code" when it happened. (formatting on save didn't do anything)

**Comment by Gys** [comment]
Actually I just now realize that formatting on save is not in the feature list....

**Comment by lukehoban** [comment]
Format on save is not implemented yet - tracking that in #14.

@mholt Could you share the full file where you see this problem?  And could you confirm whether you see this even when you save the file manually before invoking fmt?  I wasn't able to repro this myself so far.

**Comment by mholt** [comment]
I got my error in this file: https://github.com/mholt/caddy/blob/e17d43b58a3b31f16e716791e16b69752f36a5ae/caddy/sigtrap_posix.go

by adding a `fmt.Println` in one of the cases of the switch.

**Comment by sanderhahn** [comment]
One snippet that goes wrong with `goreturns` and also `goimports` is (where | is the cursor):

```go
package main

func main() {
	fmt.Println("Hello World! :)")
}
|
```

This results into:

```go
package mainimport "fmt"func main() {
	fmt.Println("Hello World! :)")
}
```

Tried to debug the `goFormat.ts`, but can't pinpoint the exact problem. If you enter another space on the last line without saving the file, the format works correctly.

**Comment by shurcooL** [comment]
I think I am seeing this issue too.

Here's a very simple reproduce case:

```Go
package main


func main() {
}
```

Note the 2 blank lines. Formatting this (with `gofmt` as the format tool) results in correctly formatting .go code.

However, consider this version with 3 blank lines:

```
package main



func main() {
}
```

Formatting this leaves 3 blank lines, which is incorrect.

**Comment by lukehoban** [comment]
Thanks for the repro cases @shurcooL  and @sanderhahn.  As far as I can tell the Go extension is requesting the right edits here, but Code isn't applying them correctly.  Following up on that in https://github.com/Microsoft/vscode/issues/144.

**Comment by shurcooL** [comment]
I've looked over `goFormat.ts` and I agree with your assessment @lukehoban. At least I didn't spot anything obviously wrong in there.

**Comment by nictuku** [comment]
Another repro case (vertical spaces to save space here):

```
package main
import (
	"log"
	"net/http"
)
func main() {
	log.Fatal(http.ListenAndServe(":8080", http.FileServer(http.Dir("/usr/share/doc"))))
}
```

Now add a spurious "fmt" import:

```
package main
import (
	"fmt"
	"log"
	"net/http"
)
func main() {
	log.Fatal(http.ListenAndServe(":8080", http.FileServer(http.Dir("/usr/share/doc"))))
}
```

When I save, it shows the red lines below "fmt", as expected. But then after Format Code, the import block becomes (note the sole double quotes)

```

import (
	"
	"log"
	"net/http"
)
```

Nice work overall, Luke and team. This is shaping up really really nicely.

**Comment by freeformz** [comment]
@nictuku FWIW: I copy-n-pasted that code, saved it as `mytest.go`, right clicked and chose "Format Code" and everything is fine.

I tested this with "goreturns" / "goimports" 
` "go.formatTool": "goreturns"` is my formatTool

**Comment by shurcooL** [comment]
> when I specify gofmt I get a missing tool prompt for some reason

I think the problem is that "do you have this tool installed" check only looks in GOPATH/bin folders (for each workspace), but not GOROOT/bin where gofmt lies.

**Comment by newhook** [comment]
Perhaps as a work around you could replace the entire document?

**Comment by tpng** [comment]
[vscode 0.10.2](https://github.com/Microsoft/vscode/issues?q=milestone%3A%22Nov+2015+-+end%22+is%3Aclosed) should have fixed this issue.

**Comment by tpng** [comment]
[vscode 0.10.2](https://github.com/Microsoft/vscode/issues?q=milestone%3A%22Nov+2015+-+end%22+is%3Aclosed) should have fixed this issue.

**Comment by lukehoban** [comment]
All - I believe this is now fixed if you update to VS Code `0.10.2` with the fix for https://github.com/Microsoft/vscode-go/issues/70.  

If anyone still sees corruption after formatting, please reopen.
